libzfs_core/_libzfs_core.py

```diff
@@ -0,0 +1,659 @@
+             using the following format :file:`{pool}/{fs}@{snap}#{tag}`
+    :rtype: list of bytes
+
+    :raises HoldReleaseFailure: if one or more existing holds could not be released.
+
+    Holds which failed to release because they didn't exist will have an entry
+    added to errlist, but will not cause an overall failure.
+
+    This call is success if ``holds`` was empty or all holds that
+    existed, were successfully removed.
+    Otherwise an exception will be raised.
+    '''
+    errlist = {}
+    holds_dict = {}
+    for snap, hold_list in holds.iteritems():
+        if not isinstance(hold_list, list):
+            raise TypeError('holds must be in a list')
+        holds_dict[snap] = {hold: None for hold in hold_list}
+    #holds_dict = {snap: {hold: None for hold in hold_list}
+    #                for snap, hold_list in holds.iteritems()}
+    with nvlist_in(holds_dict) as nvlist:
+        with nvlist_out(errlist) as errlist_nvlist:
+            ret = _lib.lzc_release(nvlist, errlist_nvlist)
+    xlate.lzc_release_xlate_errors(ret, errlist, holds)
+    # If there is no error (no exception raised by _handleErrList), but errlist
+    # is not empty, then it contains missing snapshots and tags.
+    assert all(x == errno.ENOENT for x in errlist.itervalues())
+    return errlist.keys()
+
+
+def lzc_get_holds(snapname):
+    '''
+    Retrieve list of *user holds* on the specified snapshot.
+
+    :param bytes snapname: the name of the snapshot.
+    :return: holds on the snapshot along with their creation times
+             in seconds since the epoch
+    :rtype: dict of bytes : int
+    '''
+    holds = {}
+    with nvlist_out(holds) as nvlist:
+        ret = _lib.lzc_get_holds(snapname, nvlist)
+    xlate.lzc_get_holds_xlate_error(ret, snapname)
+    return holds
+
+
+def lzc_send(snapname, fromsnap, fd, flags = []):
+    '''
+    Generate a zfs send stream for the specified snapshot and write it to
+    the specified file descriptor.
+
+    :param bytes snapname: the name of the snapshot to send.
+    :param fromsnap: if not None the name of the starting snapshot
+                     for the incremental stream.
+    :type fromsnap: bytes or None
+    :param int fd: the file descriptor to write the send stream to.
+    :param flags: the flags that control what enhanced features can be used
+                  in the stream.
+    :type flags: list of bytes
+
+    :raises SnapshotNotFound: if either the starting snapshot is not `None` and does not exist,
+                              or if the ending snapshot does not exist.
+    :raises NameInvalid: if the name of either snapshot is invalid.
+    :raises NameTooLong: if the name of either snapshot is too long.
+    :raises SnapshotMismatch: if ``fromsnap`` is not an ancestor snapshot of ``snapname``.
+    :raises PoolsDiffer: if the snapshots belong to different pools.
+    :raises IOError: if an input / output error occurs while writing to ``fd``.
```

Comment:
From this, I can infer that this function call won't return until the stream has been completely written to the file descriptor.  It might be nice to explicitly state this in the docstring.